networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/libp2p/gossip/GossipHandler.java

```diff
@@ -13,112 +13,80 @@
 
 package tech.pegasys.artemis.networking.p2p.libp2p.gossip;
 
-import com.google.common.annotations.VisibleForTesting;
-import com.google.common.eventbus.EventBus;
 import io.libp2p.core.pubsub.MessageApi;
 import io.libp2p.core.pubsub.PubsubPublisherApi;
 import io.libp2p.core.pubsub.Topic;
 import io.netty.buffer.Unpooled;
 import java.util.Set;
-import java.util.function.Consumer;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import org.apache.tuweni.bytes.Bytes;
-import org.apache.tuweni.ssz.SSZException;
-import tech.pegasys.artemis.datastructures.util.SimpleOffsetSerializer;
+import tech.pegasys.artemis.networking.p2p.gossip.TopicHandler;
 import tech.pegasys.artemis.util.collections.LimitedSet;
 import tech.pegasys.artemis.util.collections.LimitedSet.Mode;
-import tech.pegasys.artemis.util.sos.SimpleOffsetSerializable;
 
-public abstract class GossipTopicHandler<T extends SimpleOffsetSerializable>
-    implements Consumer<MessageApi> {
+public class GossipHandler implements Function<MessageApi, CompletableFuture<Boolean>> {
   private static final Logger LOG = LogManager.getLogger();
+
+  private static CompletableFuture<Boolean> VALIDATION_FAILED =
+      CompletableFuture.completedFuture(false);
+  private static CompletableFuture<Boolean> VALIDATION_SUCCEEDED =
+      CompletableFuture.completedFuture(true);
   static final int GOSSIP_MAX_SIZE = 1048576;
   private static final int MAX_SENT_MESSAGES = 2048;
 
+  private final Topic topic;
   private final PubsubPublisherApi publisher;
-  private final EventBus eventBus;
-
+  private final TopicHandler handler;
   private final Set<Bytes> processedMessages =
       LimitedSet.create(MAX_SENT_MESSAGES, Mode.DROP_LEAST_RECENTLY_ACCESSED);
 
-  protected GossipTopicHandler(final PubsubPublisherApi publisher, final EventBus eventBus) {
+  public GossipHandler(
+      final Topic topic, final PubsubPublisherApi publisher, final TopicHandler handler) {
+    this.topic = topic;
     this.publisher = publisher;
-    this.eventBus = eventBus;
+    this.handler = handler;
   }
 
-  public abstract Topic getTopic();
-
   @Override
-  public final void accept(MessageApi message) {
+  public CompletableFuture<Boolean> apply(final MessageApi message) {
     Bytes bytes = Bytes.wrapByteBuf(message.getData()).copy();
```

Comment:
Existing problem, but we shouldn't copy the data until at least after we validate the length. Ideally we wouldn't copy it if it's already in `processedMessages` but that makes the atomic add difficult.